src/metabase/query_processor/middleware/add_implicit_clauses.clj

```diff
@@ -117,14 +117,49 @@
 ;;; |                                        Add Implicit Breakout Order Bys                                         |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
+(defn- fix-order-by-field-refs
+  "This function transforms top level integer field refs in order by to corresponding string field refs from breakout
+  if present.
+
+  ## Context
+  In current situation, ie. model as a source, then aggregation and breakout, and finally order by a breakout field,
+  [[metabase.lib.order-by/orderable-columns]] returns field ref with integer id, while reference to same field, but
+  with string id is present in breakout. Then, [[add-implicit-breakout-order-by]] adds the string ref to order by.
+
+  Resulting query would contain both references, while integral is transformed differently -- it contains no casting.
+  As that is not part of group by, the query would fail.
+
+  Reference: https://github.com/metabase/metabase/issues/44653."
+  [{:keys [breakout order-by] :as inner-query}]
+  (if (or (empty? breakout) (empty? order-by))
+    inner-query
+    (let [name->breakout (into {}
+                               (keep (fn [[tag id-or-name :as clause]]
+                                       (when (and (= :field tag)
+                                                  (string? id-or-name))
+                                         [id-or-name clause])))
```

Comment:
Keying on the `id-or-name` might cause collisions, eg. if we breakout on the same field twice with different time granularity. (That's not really supported right now, but the point stands.)

However I don't know if that really matters in this case, since the ordering will still be the same.
I've noted this PR as a case to test in the multiple-breakouts doc.

Reply:
> That's not really supported right now, but the point stands.
That's true. I'm not sure, but I'd expect support for that being added as part of refs revamp. Then, hopefully, this function will become redundant.